### PR TITLE
Docker to Machine migration doc

### PR DESCRIPTION
**Pull request by gmemstr**
# Description

Adds some documentation around the steps required to move from a Docker executor to machine, and vice versa.

# Reasons

Occasionally, the machine executor is recommended to customers who require more memory for their project, but there is no clear documentation regarding the differences between the two executors, which can make the move difficult. This also lays out a list of manually installed software within the machine images.

### CODE REVIEW

**Comment by rosieyohannan** [comment]
This looks like a great addition @gmemstr ! Thanks so much - I will review today :-)

File: jekyll/_cci2/docker-to-machine.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+This document contains some general guidelines and considerations to make when
+moving from the Docker executor to machine, or vice versa.
+
+* TOC
+{:toc}
+
+## Overview
+{:.no_toc}
+
+Occiasonally, the Docker executor isn't quite the right fit for your builds -
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
The Docker executor is not the right fit for all pipelines. 
```

File: jekyll/_cci2/docker-to-machine.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+moving from the Docker executor to machine, or vice versa.
+
+* TOC
+{:toc}
+
+## Overview
+{:.no_toc}
+
+Occiasonally, the Docker executor isn't quite the right fit for your builds - 
+this can include a lack of memory or requiring more dedicated CPU  power. Moving
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
Reasons for choosing an alternative executor could include requiring more memory or more dedicated CPU  power. Moving
```

Comment (rosieyohannan) [reply jekyll/_cci2/docker-to-machine.md]: The formatting is going to go all wrong here because the paragraph is split across multiple lines.

File: jekyll/_cci2/docker-to-machine.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+to a dedicated virtual machine can help alleviate some of these  issues, but
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
to a dedicated virtual machine can help alleviate some of these issues, but 
```

File: jekyll/_cci2/docker-to-machine.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+* TOC
+{:toc}
+
+## Overview
+{:.no_toc}
+
+Occiasonally, the Docker executor isn't quite the right fit for your builds - 
+this can include a lack of memory or requiring more dedicated CPU  power. Moving
+to a dedicated virtual machine can help alleviate some of these  issues, but 
+changing out the executor isn't as easy as replacing a few lines of
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
changing out an executor isn't as easy as replacing a few lines of  
```

File: jekyll/_cci2/docker-to-machine.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+## Overview
+{:.no_toc}
+
+Occiasonally, the Docker executor isn't quite the right fit for your builds - 
+this can include a lack of memory or requiring more dedicated CPU  power. Moving
+to a dedicated virtual machine can help alleviate some of these  issues, but 
+changing out the executor isn't as easy as replacing a few lines of  
+configuration. There are some other considerations to make, such as the tools 
+and libraries required to be installed for the application and tests.
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
and libraries required to be installed for your application and tests. 
```

File: jekyll/_cci2/docker-to-machine.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+configuration. There are some other considerations to make, such as the tools 
+and libraries required to be installed for the application and tests. 
+
+## Pre-installed software
+
+By default, the machine executor images come installed with useful utilities,
+but application specific requirements will need to be installed - if the
+dependency is not installed within Ubuntu 16.04 by default or is not found on
+this list, it will need to be manually installed (note the most up to date
+list can be found here: https://raw.githubusercontent.com/circleci/image-builder/picard-vm-image/provision.sh):
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
list can be found [here](https://raw.githubusercontent.com/circleci/image-builder/picard-vm-image/provision.sh)):
```

File: jekyll/_cci2/docker-to-machine.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+* rvm
+* clojure
+* scala
+* docker
+* socat
+* nsenter
+
+\* global default
+
+Additional packages can be installed with `sudo apt-get install <package>`. IF
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
Additional packages can be installed with `sudo apt-get install <package>`. If
```

File: jekyll/_cci2/docker-to-machine.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+the package in question is not found, `sudo apt-get update` may be required 
+before installing it.
+
+## Can I run Docker containers on machine?
+
+Yes! Machine executors come installed with Docker, which could be utilized to
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
Yes! Machine executors come installed with Docker, which can be utilized to
```

File: jekyll/_cci2/docker-to-machine.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+run your applications within the container rather than installing additional
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
run your application within a container rather than installing additional
```

File: jekyll/_cci2/docker-to-machine.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+dependencies. Note it's recommended this is done with a customer Docker image
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
dependencies. Note, it is recommended this is done with a customer Docker image
```

File: jekyll/_cci2/docker-to-machine.md [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+rather than one of the convenience images, as the convenience images are built
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
rather than a CircleCI convenience image, which are built
```

File: jekyll/_cci2/docker-to-machine.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+under the assumption they will be used with the Docker executor and may be 
+tricky to work around. Since it's a dedicated virtual machine, commands to run
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
tricky to work around. Since each machine executor environment is a dedicated virtual machine, commands to run
```

File: jekyll/_cci2/docker-to-machine.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+background containers can be used normally as well.
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
background containers can be used as normal as well. 
```

File: jekyll/_cci2/docker-to-machine.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+Yes! Machine executors come installed with Docker, which could be utilized to
+run your applications within the container rather than installing additional
+dependencies. Note it's recommended this is done with a customer Docker image
+rather than one of the convenience images, as the convenience images are built
+under the assumption they will be used with the Docker executor and may be 
+tricky to work around. Since it's a dedicated virtual machine, commands to run
+background containers can be used normally as well. 
+
+Note that if you have Docker Layer Caching enabled for your account, machine
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
**Note:** if you have Docker Layer Caching enabled for your account, machine
```

File: jekyll/_cci2/docker-to-machine.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+executors can utilize this to cache your image layers for subsequent runs as
+well.
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion

```

File: jekyll/_cci2/docker-to-machine.md [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+Yes! Machine executors come installed with Docker, which could be utilized to
+run your applications within the container rather than installing additional
+dependencies. Note it's recommended this is done with a customer Docker image
+rather than one of the convenience images, as the convenience images are built
+under the assumption they will be used with the Docker executor and may be 
+tricky to work around. Since it's a dedicated virtual machine, commands to run
+background containers can be used normally as well. 
+
+Note that if you have Docker Layer Caching enabled for your account, machine
+executors can utilize this to cache your image layers for subsequent runs as
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
executors can utilize this to cache your image layers for subsequent runs.
```

File: jekyll/_cci2/docker-to-machine.md (hunk 14 above)
Comment (rosieyohannan) [reply jekyll/_cci2/docker-to-machine.md]: Just adding DLC because the abbreviation is used below

File: jekyll/_cci2/docker-to-machine.md [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+### Why use Docker executors at all?
+
+While machine executors do offer twice the memory and a more isolated 
+enviornment, there is some additional overhead regarding spin up time, and,
+depending on the approach taken for running the application, more time is taken
+to install the required dependencies or pull your Docker image. The Docker
+executor will also cache as many layers as possible from your image during
+spin-up, as opposed to the machine executor, where DLC will need to be enabled.
+
+With that in mind, there are tradeoffs to using either executor. Which one fits
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]: Maybe rephrase with a more positive tone - "All executors have pros and cons, some of which are presented here to help you decide which is right for your use case." or something?

**Comment by rosieyohannan** [comment]
Would be great to add a link to this docs from our other executor type focussed docs to help people find it. Thanks!

**Comment by gmemstr** [comment]
Appreciate the edits - going through this soon :)

File: jekyll/_cci2/docker-to-machine.md [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+make when moving from the Docker executor to machine, or vice versa.
+
+-   TOC {:toc}
+
+Overview
+--------
+
+{:.no\_toc}
+
+Occiasonally, the Docker executor isn't quite the right fit for your
```
Comment (rosieyohannan) [inline jekyll/_cci2/docker-to-machine.md]:
```suggestion
Occasionally, the Docker executor isn't quite the right fit for your
```

**Comment by rosieyohannan** [comment]
@gmemstr Hey! Sorry it's taken me so long to get back to this. I just pushed some changes to fix formatting and bring some things in line with our style guide. Also I linked to this new doc from the executor types overview so people can find it. If you can think of other useful places to link to it from go ahead! Thanks so much for the new doc :-)
